parquet-hadoop/src/main/java/org/apache/parquet/hadoop/DirectCodecFactory.java

```diff
@@ -0,0 +1,364 @@
+import org.xerial.snappy.Snappy;
+
+import org.apache.parquet.bytes.ByteBufferAllocator;
+import org.apache.parquet.bytes.BytesInput;
+import org.apache.parquet.hadoop.CodecFactory.BytesCompressor;
+import org.apache.parquet.hadoop.HeapCodecFactory.HeapBytesCompressor;
+import org.apache.parquet.hadoop.HeapCodecFactory.HeapBytesDecompressor;
+import org.apache.parquet.hadoop.metadata.CompressionCodecName;
+
+/**
+ * Factory to produce compressors and decompressors that operate on java
+ * direct memory, without requiring a copy into heap memory (where possible).
+ */
+public class DirectCodecFactory extends CodecFactory<BytesCompressor, DirectBytesDecompressor> implements AutoCloseable {
+//  private static final Log LOG = Log.getLog(DirectCodecFactory.class);
+
+  private final ByteBufferAllocator allocator;
+
+  public DirectCodecFactory(Configuration config, ByteBufferAllocator allocator) {
+    super(config);
+    Preconditions.checkNotNull(allocator, "ByreBufferAllocator cannot be null.");
+    this.allocator = allocator;
+  }
+
+  private ByteBuffer ensure(ByteBuffer buffer, int size) {
+    if (buffer == null) {
+      buffer = allocator.allocate(size);
+    } else if (buffer.capacity() >= size) {
+      buffer.clear();
+    } else {
+      allocator.release(buffer);
+      release(buffer);
+      buffer = allocator.allocate(size);
+    }
+    return buffer;
+  }
+
+  ByteBuffer release(ByteBuffer buffer) {
+    if (buffer != null) {
+      allocator.release(buffer);
+    }
+    return null;
+  }
+
+  @Override
+  protected BytesCompressor createCompressor(final CompressionCodecName codecName, final CompressionCodec codec,
+                                             int pageSize) {
+
+    if (codec == null) {
+      return new NoopCompressor();
+    } else if (codecName == CompressionCodecName.SNAPPY) {
+      // avoid using the Parquet Snappy codec since it allocates direct buffers at awkward spots.
+      return new SnappyCompressor();
+    } else {
+
+      // todo: move zlib above since it also generates allocateDirect calls.
+      return new HeapBytesCompressor(codecName, codec, pageSize);
+    }
+  }
+
+  @Override
+  protected DirectBytesDecompressor createDecompressor(final CompressionCodec codec) {
+    // This is here so that debugging can be done if we see inconsistencies between our decompression and upstream
+    // decompression.
+    // if (true) {
+    // return new HeapFakeDirect(codec);
+    // }
+
+    if (codec == null) {
+      return new NoopDecompressor();
+    } else if (DirectCodecPool.INSTANCE.codec(codec).supportsDirectDecompression()) {
+      return new FullDirectDecompressor(codec);
+    } else {
+      return new IndirectDecompressor(codec);
+    }
+  }
+
+  public void close() {
+    release();
+  }
+
+  /**
+   * Keeping this here for future debugging versus using custom implementations below.
+   */
+  private class HeapFakeDirect extends DirectBytesDecompressor {
```

Comment:
Currently unused, we should just have good enough tests between the direct and heap versions that we don't need this stub to debug between the two.